File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithWrappedDependency.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.application.graph.All;
import ru.tinkoff.kora.application.graph.ValueOf;
import ru.tinkoff.kora.application.graph.Wrapped;
import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

@KoraApp
public interface AppWithWrappedDependency {
    @Root
    default Class2 class2(Class1 class1) {
        return new Class2();
    }

    @Root
    default Class3 class3(ValueOf<Class1> class1) {
        return new Class3();
    }

    @Root
    default Class4 class4(All<ValueOf<Class1>> class1) {
        return new Class4();
    }

    @Root
    default Class2 class2ValueWrapped(Wrapped<Class1> class1) {
        return new Class2();
    }

    @Root
    default Class3 class3Wrapped(ValueOf<Wrapped<Class1>> class1) {
        return new Class3();
    }

    @Root
    default Class4 class4Wrapped(All<ValueOf<Wrapped<Class1>>> class1) {
        return new Class4();
    }


    default Wrapped<Class1> class1() {
        var c1 = new Class1();
        return () -> c1;
    }

    class Class1 {}

    class Class2 {}

    class Class3 {}

    class Class4 {}
}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithWrappedDependency.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.application.graph.All
import ru.tinkoff.kora.application.graph.ValueOf
import ru.tinkoff.kora.application.graph.Wrapped
import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root

@KoraApp
interface AppWithWrappedDependency {
    @Root
    fun class2(class1: Class1): Class2 {
        return Class2()
    }

    @Root
    fun class3(class1: ValueOf<Class1>): Class3 {
        return Class3()
    }

    @Root
    fun class4(class1: All<ValueOf<Class1>>): Class4 {
        return Class4()
    }

    @Root
    fun class2ValueWrapped(class1: Wrapped<Class1>): Class2 {
        return Class2()
    }

    @Root
    fun class3Wrapped(class1: ValueOf<Wrapped<Class1>>): Class3 {
        return Class3()
    }

    @Root
    fun class4Wrapped(class1: All<ValueOf<Wrapped<Class1>>>): Class4 {
        return Class4()
    }

    fun class1(): Wrapped<Class1> {
        val c1 = Class1()
        return Wrapped { c1 }
    }

    class Class1
    class Class2
    class Class3
    class Class4
}
